egps/egp-15/egp-15-retro.ts

```diff
@@ -0,0 +1,121 @@
+
+// Helpers
+import * as addresses from '../helpers/addresses'
+import { createCallHash } from '../helpers/hashing'
+
+async function proposal() {
+  // Setup your signer
+  // Using a ledger, connect to metamask
+  const providerUrl = ''
+  const provider = new ethers.providers.JsonRpcProvider(providerUrl)
+  const signer = new LedgerSigner(provider, undefined, "m/44'/60'/3'/0/0")
+
+  // Setup your interfaces
+  const timelockInterface = new ethers.utils.Interface(timelockData.abi) // TimeLock is like sudo, you always need it.
+  const treasuryInterface = new ethers.utils.Interface(treasuryData.abi) // TimeLock is like sudo, you always need it.
+  const lockingVaultInterface = new ethers.utils.Interface(lockingVaultData.abi)
+
+  // New Params
+  // @TODO - fill in.
+
+  // Connect the signer to the coreVotingContract, this is where your proposals will fed into.
+  const coreVoting = CoreVoting__factory.connect(addresses.CoreVoting, signer)
+
+  // --- main egp logic ---
+  /**
+   * The Timelock contract has to call deposit() on the locking vault contract through the Treasury contract...
+   * 1. Generate calldata for deposit() call on locking vault
+   * 2. Generate calldata for genericCall() call on treasury
+   * 3. Generate callhash for timelock
+   */
+
+  /** Generate calldata for locking vault deposit */
+  const retroAmount = ethers.utils.parseEther('19531.25')
+  const v1StartAmount = ethers.utils.parseEther('54725.06')
+  const lockingVaultCalldata = lockingVaultInterface.encodeFunctionData('deposit', [
+    addresses.ComponentMultisig,
+    retroAmount.add(v1StartAmount),
+    addresses.ComponentDelegate,
+  ])
+
+  /** Generate calldata for treasury genericcall */
+  const callData = treasuryInterface.encodeFunctionData('genericCall', [
+    addresses.FrozenLockingVaultProxy,
+    lockingVaultCalldata,
+  ])
+
+  /**
+   * Take the calldata and convert it to the callhash.
+   * Param BytesLike[] - An arrary of encoded calldata
+   * Param string[] - An array of addresses, index must match that from the first parameter
+   * Notes:
+   * - You can pass in as many "chained calls" as you like, just match the calldata to the addresses in the two parameters.
+   */
+  const callHash = await createCallHash(
+    [callData], // calldata
+    [addresses.Treasury] // Target treasury to move funds
+  )
+
+  /**
+   * Encode proposal to be sent to the core voting contract
+   * Notes:
+   * - Simple add the multiple call hashes into the array.
+   * - You can have multiple callHashes, if a proposal affects multiple different vaults or other DAO contracts.
+   * - eg: Proposal updates a voting threshold, and moves funds from the treasury. This would be two callHashes.
+   */
+  const calldataCv = timelockInterface.encodeFunctionData('registerCall', [callHash])
+
+  /**
+   * Creates the expiery for the proposal, no need to modify.
+   * Default: xxxx
+   */
+  const expiryDate = 15378000 // TODO change to something automatic.
```

Comment:
Perhaps something like:
```
getExpiryDate() {
  const secondsPerDay = 86400;
  const blocksPerDay = secondsPerDay / 15;
  const daysOpen = 7;
  return provider.getBlock() + blocksPerDay * daysOpen;
}
```

Reply:
good catch that this needs to be a block number! [contract comments say timestamp](https://github.com/element-fi/council/blob/main/contracts/CoreVoting.sol#L129)